contribs/gnofaucet/gh.go

```diff
@@ -0,0 +1,106 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/google/go-github/v64/github"
+)
+
+// getGithubMiddleware sets up authentication middleware for GitHub OAuth.
+// If clientID and secret are empty, the middleware does nothing.
+//
+// Parameters:
+// - clientID: The OAuth client ID issued by GitHub when registering the application.
+// - secret: The OAuth client secret used to securely authenticate API requests.
+// - cooldown: A cooldown duration to prevent several claims from the same user.
+//
+// GitHub OAuth applications require a client ID and secret to authenticate users securely.
+// These credentials are obtained when registering an application on GitHub at:
+// https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authenticating-to-the-rest-api-with-an-oauth-app#registering-your-app
+func getGithubMiddleware(clientID, secret string, coolDownLimiter *CooldownLimiter) func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(
+			func(w http.ResponseWriter, r *http.Request) {
+				// github Oauth flow is enabled
+				if secret == "" || clientID == "" {
+					// Continue with serving the faucet request
+					next.ServeHTTP(w, r)
+
+					return
+				}
+
+				// Extracts the authorization code returned by the GitHub OAuth flow.
+				//
+				// When a user successfully authenticates via GitHub OAuth, GitHub redirects them
+				// to the registered callback URL with a `code` query parameter. This code is then
+				// exchanged for an access token.
+				//
+				// Reference: https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps#2-users-are-redirected-back-to-your-site-by-github
+				code := r.URL.Query().Get("code")
+				if code == "" {
+					http.Error(w, "missing code", http.StatusBadRequest)
+					return
+				}
+
+				user, err := exchangeCodeForUser(r.Context(), secret, clientID, code)
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusBadRequest)
+					return
+				}
+
+				// Just check if given account have asked for faucet before the cooldown period
+				allowedToClaim, err := coolDownLimiter.CheckCooldown(user.GetLogin())
+				if err != nil {
+					http.Error(w, err.Error(), http.StatusInternalServerError)
+					return
+				}
+
+				if !allowedToClaim {
+					http.Error(w, "user is on cooldown", http.StatusTooManyRequests)
+					return
+				}
+
+				// Possibility to have more conditions like accountAge, commits, pullRequest, etc.
```

Comment:
There needs to be a maximum amount of gnot the user can request from the faucet, ever

Reply:
Implemented a max-claimable-limit here 703b16f92a45ebb4bbd19d6e5ccdb4406b37d1e4